ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldStateDownloader.java

```diff
@@ -178,28 +205,21 @@ private void requestDataFromPeer(final BlockHeader header, final EthPeer peer) {
     // Collect data to be requested
     final List<Task<NodeDataRequest>> toRequest = getTasksForNextRequest();
 
+    if (toRequest.isEmpty()) {
+      requestNodeData(header);
+      return;
+    }
+
     // Request and process node data
     sendAndProcessRequests(peer, toRequest, header)
         .whenComplete(
             (task, error) -> {
-              final boolean done;
-              synchronized (this) {
-                outstandingRequests.remove(task);
-                done =
-                    status == Status.RUNNING
-                        && outstandingRequests.size() == 0
-                        && pendingRequests.allTasksCompleted();
-              }
-              if (done) {
-                // We're done
-                final Updater updater = worldStateStorage.updater();
-                updater.putAccountStateTrieNode(header.getStateRoot(), rootNode);
-                updater.commit();
-                markDone();
-              } else {
-                // Send out additional requests
-                requestNodeData(header);
+              if (error != null
+                  && !(ExceptionUtils.rootCause(error) instanceof RejectedExecutionException)) {
+                LOG.error("World state data request failed", error);
```

Comment:
consider adding some context to the log message to make debugging easier (eg peer, toRequest, header, etc)

Reply:
Again there's no useful context we can include here beyond the exception.  This is something completely unexpected that happened somewhere through the entire create, send and process request.  It shouldn't ever happen but if it does we want to see the stack trace. Most errors are already handled in the details - this is the equivalent of the uncaught exception handler.

Reply:
If we had what eric asked for at debug or trace it would help to re-create a reproducible state.

Reply:
What log message are you actually expecting to see here? I've been working in this code a fair bit and can't imagine what we could add here that would actually be useful.